### PR TITLE
Move spinner sample point to the end time on the timeline

### LINKED ISSUE #29338: Spinner sample point should be at the end time on the timeline
### Type

Game behaviour

### Bug description

Spinners play their hitsound at the end of the object, so it would make sense if the sample point to edit the hitsound is also at the end of the object on the timeline. With the current behaviour you're forced to navigate to the start of a long spinner in order to change the sound at the end which is a nuisance.

### Screenshots or videos

osu!_jvufnmHmrH

### Version

2024.731.0-lazer

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/16532752/compressed-logs.zip)

### PR DESCRIPTION
**Pull request by OliBomby**
closes #29338

I had to add an interface to prevent this timeline change also applying to taiko rolls, ctb banana showers, and mania hold notes.

### CODE REVIEW

File: osu.Game/Rulesets/Objects/Types/IHasHitsoundTiming.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    /// A HitObject which has its hitsound at a specific time along its duration. Will be used for editor timeline display.
+    /// </summary>
+    public interface IHasHitsoundTiming
+    {
+        /// <summary>
+        /// The current hitsound timing of this hit object.
+        /// 0 means at the start of the object, 1 means at the end of the object.
+        /// </summary>
+        float HitsoundTiming { get; }
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/Types/IHasHitsoundTiming.cs:15]: This in general being [0;1]-ranged is a bit baffling too. It should be an absolute time instant IMO (so `.GetEndTime()` in this case).

**Review by bdach — COMMENTED** [review]
I'm not super sure whether this makes sense, to be completely fair.

First of all, the `Samples` of the spinner also affect `spinnerspin`, for better or worse:

https://github.com/ppy/osu/blob/24a0a3c47fd7ed1e12848b80e4882289df23fb2a/osu.Game.Rulesets.Osu/Objects/Spinner.cs#L104-L115

While `spinnerspin` doesn't have bank, what it does have is *volume*. So it feels misleading that changing a sample piece at the *end* of the spinner will change the volume of spin.

Secondly, the entire API design is a bit weird. The structure is such that `Samples` *generally* play at the *end* time of the object (because that's when it usually becomes hit). So I'd expect the sample point to be *at the end by default* for objects that have duration, and not *the start*. The fact that it's not the case for taiko/catch/mania is basically a bunch of coincidences at once (one being that the taiko drum/mania keys make the noises rather than the individual objects, and that catch juice streams/banana showers aren't objects on their own so much as *collections* of ones).

**Comment by OliBomby** [comment]
> While `spinnerspin` doesn't have bank, what it does have is _volume_. So it feels misleading that changing a sample piece at the _end_ of the spinner will change the volume of spin.

Yeah, I think having the sample piece for spinner spin volume at the end or even the start is kinda misleading. Ideally you want a separate sample piece for the spinner spin sample, similar to how sliders currently work, but that's out of scope for this PR. Hitsounders mostly care about the hitsound at the end for spinners, so I'd say this is overal an improvement.

> Secondly, the entire API design is a bit weird. The structure is such that `Samples` _generally_ play at the _end_ time of the object (because that's when it usually becomes hit). So I'd expect the sample point to be _at the end by default_ for objects that have duration, and not _the start_. The fact that it's not the case for taiko/catch/mania is basically a bunch of coincidences at once (one being that the taiko drum/mania keys make the noises rather than the individual objects, and that catch juice streams/banana showers aren't objects on their own so much as _collections_ of ones).

I hadn't thought about this until you mentioned it, but the `Samples` generally playing at the end time of the object is such a weird behaviour. Generally in instruments and rhythm games you have no sounds on the release of a keypress. Spinners feel like the exception to me.

File: osu.Game/Rulesets/Objects/Types/IHasHitsoundTiming.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+namespace osu.Game.Rulesets.Objects.Types
+{
+    /// <summary>
+    /// A HitObject which has its hitsound at a specific time along its duration. Will be used for editor timeline display.
+    /// </summary>
+    public interface IHasHitsoundTiming : IHasDuration
+    {
+        /// <summary>
+        /// The current hitsound timing of this hit object.
+        /// 0 means at the start of the object, 1 means at the end of the object.
```
Comment (bdach) [inline osu.Game/Rulesets/Objects/Types/IHasHitsoundTiming.cs:13]: outdated xmldoc

**Review by bdach — COMMENTED** [review]
Checks failing.

Would want @ppy/team-client opinions on this in general as well.

File: osu.Game/Rulesets/Objects/Types/IHasHitsoundTiming.cs (hunk 1 above)
Comment (OliBomby) [reply osu.Game/Rulesets/Objects/Types/IHasHitsoundTiming.cs]: I kinda dislike how complex the logic has gotten after your suggestion, but it does make sense to not over-specialise it to the `TimelineHitObjectBlueprint` usecase.

**Comment by smoogipoo** [comment]
I'm quite confused by this...

In my mind samples are assumed to be at end time by default. This is because in general hitobjects aren't hit until their nested hitobjects have all been judged, and those are assumed to span the entire duration of the hitobject.  

All current implementations of hitobjects with duration have some form of:
```
void CheckForResult()
{
    if (Time < EndTime)
        return;
}
```

https://github.com/ppy/osu/blob/dd675fef45db1693db4087f088f99a62764021b7/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs#L295-L296
https://github.com/ppy/osu/blob/dd675fef45db1693db4087f088f99a62764021b7/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSpinner.cs#L254-L255

Same thing, different way of writing:

https://github.com/ppy/osu/blob/dd675fef45db1693db4087f088f99a62764021b7/osu.Game.Rulesets.Taiko/Objects/Drawables/DrawableDrumRoll.cs#L143-L144
https://github.com/ppy/osu/blob/dd675fef45db1693db4087f088f99a62764021b7/osu.Game.Rulesets.Catch/Objects/Drawables/DrawableCatchHitObject.cs#L65-L66

Mania differs because it doesn't have a sample on the hold note itself.

It's not a hard-line guarantee, but I'd say it's a common implementation. Reading more above, it seems I'm not the only one that independently came to this conclusion... I don't foresee this being used outside of the editor, if it even needs to exist in the first place, so I would much prefer it being specialised even to the point of specifying it's editor-specific in the interface name. In that case, I would also prefer `[0;1]` range rather than absolute time.

**Comment by OliBomby** [comment]
> It's not a hard-line guarantee, but I'd say it's a common implementation. Reading more above, it seems I'm not the only one that independently came to this conclusion... I don't foresee this being used outside of the editor, if it even needs to exist in the first place, so I would much prefer it being specialised even to the point of specifying it's editor-specific in the interface name. In that case, I would also prefer `[0;1]` range rather than absolute time.

So I see two possibilities: Either I revert to the `[0;1]` range, or I remove the interface entirely and accept that taiko rolls, ctb banana showers, and mania hold notes all should have their `Samples` at the end time.

**Comment by bdach** [comment]
> I don't foresee this being used outside of the editor, if it even needs to exist in the first place, so I would much prefer it being specialised even to the point of specifying it's editor-specific in the interface name. In that case, I would also prefer `[0;1]` range rather than absolute time.

This is a bit of a non-sequitur to me (why does editor usage impact whether this needs to be [0;1] range or not?) but ok, I guess I could live with it. I just feel like absolute time is a nicer thing to use in this context given everything else on hitobjects uses absolute time.

I dunno, I have no strong opinions, but if I were to lean in a specific direction then I feel like this PR could just be closed to no harm in general because as stated above even for osu! spinners the sample point being at the start of the object happens to make sense anyway because of `spinnerspin`.

**Comment by peppy** [comment]
I think this might be best left until we have the sample editor setup, but even so we'll likely need a similar interface.

But is there any valid case for putting a sample anywhere but the *start or end*? If not, I'd prefer this is just a `bool ShowSampleAtEndTime`.

As mentioned by @bdach, the sample settings also affects the spinner sound, so I dunno how valid this is. I don't think many people are going to care that it shows at the start so maybe we should leave it for now (until we have a sample editor where we want to show samples *precisely* when they happen so it matches visually with what you hear when using the editor).

**Comment by bdach** [comment]
> But is there any valid case for putting a sample anywhere but the _start or end_ (I can't think of one)? If not, I'd prefer this is just a `bool ShowSampleAtEndTime`.

I really don't like that.

Let me just emphasise one thing: we're adding a thing to a *hitobject class*. A thing that more or less amounts to *controlling visual layout of editor*, no less. If this were an actual absolute time instant of sample playback, I'd accept it, but doesn't this living in a *hitobject class* feel like a blatant break of layering? That stuff should be nowhere near `Spinner`.

**Comment by peppy** [comment]
Yeah sure, I agree with that. It should be something in the ruleset API which allows customising the display.

**Comment by smoogipoo** [comment]
I'm going to go ahead and close this as we seemingly all disagree with it.

**Comment by bdach** [comment]
> But we don't currently have a `Ruleset.CreateTimeline()` or whatever.

I think we'll need it eventually, so make of that what you will.

**Pull request closed without merging**